### PR TITLE
update Janet ecosystem; janet: update to 1.22.0; add new port: jpm

### PR DESCRIPTION
**Pull request by tsujp**
###### Tested on

macOS 12.1 21C52 arm64
Xcode 13.2.1 13C100

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] ~referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?~ **(none applicable).**
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [ ] ~tried a full install with `sudo port -vst install`?~ **<-- I use an already rootless MacPorts installation and this command wants to change to user `macports` which doesn't exist and I don't really want to make one.**
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by tsujp** [comment]
I've added a `notes` section on `jpm` because I intend to package `jpm` for MacPorts (I'm doing it now). I think `jpm` should be provided with default installation options or configuration so that it installs packages into directories managed by MacPorts instead of manually maintaining all Janet packages. For instance, we do not do with this `npm` for Nodejs.

Current `jpm` packages: https://github.com/macports/macports-ports/pull/10793

**Comment by tsujp** [comment]
Actually no x2 I'm going back to the original plan for a few reasons:

1. `jpm` is no longer installed by default with Janet (as mentioned before), and because of that...
2. The current Janet ecosystem (i.e. https://github.com/macports/macports-ports/pull/10793) breaks because it assumes `jpm` is installed (see note ^1 below).
3. Adding a `jpm` package into MacPorts configured to use the prefix of the MacPorts installation for Janet results in those files being tracked _and_ package management for every single package in the Janet ecosystem not being duplicated into MacPorts (see note ^2 below).

**Note ^1:**

This means we either need to package `jpm` into MacPorts so that the existing Janet ecosystem can work with an updated Janet (this PR) _or_ encourage users to install `jpm` outside of MacPorts' purview. I don't like suggesting the latter.


**Note ^2:**

This is a benefit I feel because as it stands `janet-uri` (the MacPorts package) with an updated Janet and jpm is broken because it uses the tag `0.0.1` release of that package. There is no later tag, so we would need to change it to a commit hash. That's fine except why duplicate the effort versus letting jpm itself (tracked by MacPorts still) install it where it works due to jpm knowing to get the latest commit.

**Comment by tsujp** [comment]
Okay I've been having a bit of a conversation with myself in this thread (that's a non-issue I just think it's funny). There was a gap in my understanding regarding "directories managed by MacPorts". MacPorts cannot remove (and there is no configurable option I know of) directories or files that are not present in the image of the port which is created from the portfile at installation time. So, unless we do heinous things like updating that image file with hooks on `jpm` itself (noooooooooooooooooooo) I now grok why https://github.com/macports/macports-ports/pull/10793 was done the way it was.

Anyway, it still stands that `jpm` is not packaged with Janet anymore (by default) so the solution there is probably to add another package `jpm` (which I already have finished). I think providing `jpm` as a separate port is good since then people who don't want or care about `jpm` can still install `janet` without the extra (for them) bloat.

Finally, what I am going to do tomorrow is update https://github.com/macports/macports-ports/pull/10793 so that it uses this `jpm` I am packaging, and also bump all those `janet-*` ports because with the update of Janet itself to 1.22.0 the use of those ports as they are will not work.

File: lang/janet/Portfile [hunk 1]
```diff
@@ -5,8 +5,8 @@ PortGroup           github 1.0
 PortGroup           legacysupport 1.1
 PortGroup           makefile 1.0
 
-github.setup        janet-lang janet 1.15.5 v
-revision            2
+github.setup        janet-lang janet 1.22.0 v
+revision            3
```
Comment (reneeotten) [inline lang/janet/Portfile:9]: the revision field should be reset to zero when updating the version

**Review by reneeotten — CHANGES_REQUESTED** [review]

File: lang/janet/Portfile [hunk 2]
```diff
@@ -23,21 +23,46 @@ long_description    Janet is a functional and imperative programming \
...
+(2) Use jpm packages via jpm itself. It is _not_ recommended\
+    to install packages globally as they would be unmanaged by\
+    MacPorts, however, jpm has functionality to install packages\
+    at a custom directory (e.g. local to your janet project).\
+    This is configured via JANET_TREE for jpm. See the\
+    installation notes for jpm for more information.
+
+    e.g.
+        \$ port install jpm
+        \$ JANET_TREE ~/projects jpm install joy
```
Comment (halostatue) [inline lang/janet/Portfile:63]: Just a drive-by comment, but this should probably be `\$ JANET_TREE=~/projects jpm install joy`.

Comment (tsujp) [reply lang/janet/Portfile:63]: Hehe missed the cheeky equals, thanks

File: lang/janet/Portfile (hunk 1 above)
Comment (tsujp) [reply lang/janet/Portfile:9]: Cheers, will fix

**Comment by tsujp** [comment]
@halostatue @reneeotten Janet Portfile cleaned up and new changes pushed.

Going to do jpm's Portfile now then the `janet/` rest of the ports.

**Comment by tsujp** [comment]
Notes 1:

By default directory`${prefix}/lib/janet` has nothing in it except for a single zero-byte hidden file `.turd_janet`.

Notes 2:

Fresh installation of `jpm` via the Portfile results in `${prefix}/lib/jpm` looking like this (all output from `tree -a`):

Fresh:

```
.
├── .manifests
│   └── jpm.jdn
└── jpm
    ├── cc.janet
    ├── cgen.janet
    ├── cli.janet
    ├── commands.janet
    ├── config.janet
    ├── dagbuild.janet
    ├── declare.janet
    ├── default-config.janet
    ├── init.janet
    ├── make-config.janet
    ├── pm.janet
    ├── rules.janet
    └── shutil.janet

2 directories, 14 files
```

Performing `port uninstall jpm` will completely remove this entire directory structure. If you install a package with `jpm` -- for example package `uri` -- it looks like this:

_the `.cache` folder has been manually deleted as this is a clone of the respective packages git repo_

```
.
├── .manifests
│   ├── jpm.jdn
│   ├── pkgs.jdn
│   └── uri.jdn
├── _uri.a
├── _uri.meta.janet
├── _uri.so
├── jpm
│   ├── cc.janet
│   ├── cgen.janet
│   ├── cli.janet
│   ├── commands.janet
│   ├── config.janet
│   ├── dagbuild.janet
│   ├── declare.janet
│   ├── default-config.janet
│   ├── init.janet
│   ├── make-config.janet
│   ├── pm.janet
│   ├── rules.janet
│   └── shutil.janet
├── pkgs.janet
└── uri.janet

2 directories, 21 files
```

Then performing `port uninstall jpm` results in the following leftover directory structure:

```
├── .manifests
│   ├── pkgs.jdn
│   └── uri.jdn
├── _uri.a
├── _uri.meta.janet
├── _uri.so
├── pkgs.janet
└── uri.janet

1 directory, 7 files
```

**Comment by tsujp** [comment]
Made a small change to janet so it can find `jpm` installed packages bethey from MacPorts (e.g. the `janet-` ones) or otherwise.

**Comment by tsujp** [comment]
Okay this is ready for review now :)

File: lang/janet/Portfile (hunk 1 above)
Comment (tsujp) [reply lang/janet/Portfile]: All good now @reneeotten ?

**Comment by tsujp** [comment]
Discussion an issue in the Janet community about `janet-spork`, when that is resolved this will be ready for review for realsies since that port is broken at the moment.

**Comment by mascguy** [comment]
Based on a quick cursory review, things are coming together. And it's clear you've put a lot of effort into this. Thank you so much!

One minor point, though: If you're updating everything that currently uses the Janet portgroup, you don't need to rename it to v1.1. Rather, just keep it named as v1.0.

**Comment by tsujp** [comment]
> Based on a quick cursory review, things are coming together. And it's clear you've put a lot of effort into this. Thank you so much!

Thank you for the review!

Yeah it's been a bit of a journey learning some more stuff, about 10 ish hours of work hehe (very good fun though) but in future much, much, much faster because of it.

> One minor point, though: If you're updating everything that currently uses the Janet portgroup, you don't need to rename it to v1.1. Rather, just keep it named as v1.0.

Ah okay. That was one thing I wasn't sure about and I gave the portgroup folder an `ls` and found some non-`1.0`s there so figured since I am changing how the portgroup works I should bump the version. I'll take it back to `1.0` then.

**Comment by tsujp** [comment]
@mascguy if you could re-review when you can that'd be great!

**Review by mascguy — APPROVED** [review]
LGTM.

@reneeotten , anyone... ?

**Comment by reneeotten** [comment]
sorry, my review was just a fly-by-comment. Much has changed since then and I have no time to review all of that.

**Comment by tsujp** [comment]
> sorry, my review was just a fly-by-comment. Much has changed since then and I have no time to review all of that.

Could you dismiss your review then or mark it resolved? It's blocking a merge (without others doing it for you) because it's unresolved.

**Comment by mascguy** [comment]
Just tested locally with all of this, and looks great. Thanks again for all of your work on this Jordan!

**Comment by tsujp** [comment]
> Just tested locally with all of this, and looks great. Thanks again for all of your work on this Jordan!

My pleasure and thank you for the review and testing effort. I've got some more stuff I am going to do on top of this. Perhaps I should apply for merge privileges or should I do more PRs like this beforehand would you say?
